### DISCUSSION TITLE
Parallel Execution of HSCS Transactions

### DISCUSSION THREAD

**Discussion by mshakeg** [hiero-ledger/hiero-improvement-proposals#677] [Ideas]
## Background
EIP-648 proposes a simple and optional mechanism for enabling the EVM to process transactions in parallel[[1]](https://github.com/ethereum/EIPs/issues/648). As of now, the EIP has not been implemented on Ethereum or any other EVM chain except potentially BSC, which has publicly disclosed its efforts in parallel transaction execution(however may have taken their own approach)[[5]](https://www.bnbchain.org/en/blog/new-milestone-the-implementation-of-parallel-evm-2-0/). However, multiple non-EVM chains such as Solana, Aptos, and Sui have implemented parallel transaction execution to increase the throughput of their smart contract services significantly. Both Sui and Solana enable this using an implementation similar to that proposed in EIP-648, i.e., either optional or mandatory specification of read and write access lists to state by a transaction[[2]](https://solana.com/news/sealevel---parallel-processing-thousands-of-smart-contracts)[[4]](https://docs.sui.io/devnet/learn/how-sui-works). On the other hand, Aptos uses a more novel mechanism called Block-STM that enables it to process 160k non-trivial Move transactions per second[[3]](https://medium.com/aptoslabs/block-stm-how-we-execute-over-160k-transactions-per-second-on-the-aptos-blockchain-3b003657e4ba).

## Proposal
This discussion proposes the implementation of parallel transaction execution for the Hedera Smart Contract Service (HSCS) in a manner similar to that proposed in EIP-648. However, unlike EIP-648, which uses RLP lists to determine the ranges of addresses a transaction can access to read and/or write, we propose specifying the exact Hedera entity ids/addresses (either account/contract/token ids). Since Hedera entity ids are not distributed pseudorandomly across the address/id space like Ethereum addresses, using RLP lists as EIP-648 proposes would not be practical, as all entity ids would fall into a single range. Therefore, specifying the entity id in the read and write dependency lists allows for transactions that share at least one entity dependency to be assigned to the same thread. Note as EIP-648 proposes this would be done on a per block level; see the **Client Implementation** algorithm in EIP-648.

To incentivize and disincentivize correct and incorrect entity ids in the read and write dependency lists, a reward and penalty mechanism is proposed. A correct entity id is an entity id that is indeed called in a given transaction's execution, and an incorrect entity id is an entity that is never called. A penalty gas is charged preemptively for each entity specified in the dependency lists, and for every entity that is called correctly, gas is awarded that is either equal to the penalty (in which case, it is an exact refund for the penalty charged) or slightly more to incentivize developers to specify correct entity dependencies. This allows for transactions to be optimally assigned to threads for parallel transaction execution, resulting in increased throughput, as opposed to developers simply not specifying dependency lists. Additionally, the mirror node can expose an endpoint to determine contracts that are invoked in the static execution of a transaction if the exact read&write lists of entity dependencies are not known in advance; similar reasoning for static `eth_call` or `eth_estimateGas`.

If a transaction calls the HTS precompile (_**0x167**_) to read and/or write from/to state, then instead of specifying the precompile address in the read and/or write entity dependency lists, the Hedera entity addresses that are being read and/or updated via the precompile contract can be specified. Since the HTS precompile contract is frequently called by many smart contracts, it makes more sense to specify the upstream entities that are being read from or written to, in order to enable parallelization across transactions that all call the HTS precompile contract but are not reading from or writing to the same entity.

Parallelization of such transactions that call the HTS precompile can be further optimised by considering dependency in HTS actions between transactions, though this will require explicit specification of the type of action and its entity dependencies. For example(**Example 1**), if Alice signs 2 contract calls to separate contracts **within the same "block"** that do a Fungible HTS transfer from her account to Bob's account but on 2 separate HTS tokens, then these 2 transactions can be executed in parallel. Or in another example(**Example 2**), if Alice signs a contract call to contract X that transfers TokenA to Bob and Carol signs a contract call to contract Y that also transfers TokenA but to Dave, then both Alice's and Carol's transactions can be executed in parallel.

#### Example 1 Write Dependency List:

Alice's 1st transaction's write dependency list:
```jsonc
[
  {
    "type": "ContractCall",
    "contract": "X" // if X is the target contract i.e. "to" then it does NOT need to be explicitly stated in the dependency list, but can be inferred from the transaction's "to" value
  },
  {
    "type": "FungibleTokenTransfer",
    "sender": "AliceId",
    "receiver": "BobId",
    "token": "TokenAId"
  }
]
```

Alice's 2nd transaction's write dependency list:

```jsonc
[
  {
    "type": "ContractCall",
    "contract": "Y"
  },
  {
    "type": "FungibleTokenTransfer",
    "sender": "AliceId",
    "receiver": "BobId",
    "token": "TokenBId"
  }
]
```

#### Example 2 Write Dependency List:

Alice's transaction's write dependency list:
```jsonc
[
  {
    "type": "ContractCall",
    "contract": "X"
  },
  {
    "type": "FungibleTokenTransfer",
    "sender": "AliceId",
    "receiver": "BobId",
    "token": "TokenAId"
  }
]
```

Carol's transaction's write dependency list:

```jsonc
[
  {
    "type": "ContractCall",
    "contract": "Y"
  },
  {
    "type": "FungibleTokenTransfer",
    "sender": "CarolId",
    "receiver": "DaveId",
    "token": "TokenAId"
  }
]
```

## Potential Issues
* The bottleneck may be disk I/O operations, not CPU processing power, so parallel transaction execution would not necessarily increase smart contract throughput; this is supposedly the reason for EIP-648 not being implemented.
* It may increase the CPU requirements for consensus nodes.

## Alternative Considerations
* Investigate whether the Block-STM approach would yield higher throughput on the EVM than the read & write dependency access lists approach.

## References

[1] [Easy parallelizability (EIP-648)](https://github.com/ethereum/EIPs/issues/648)
[2] [Sealevel — Parallel Processing Thousands of Smart Contracts](https://solana.com/news/sealevel---parallel-processing-thousands-of-smart-contracts)
[3] [Block-STM: How We Execute Over 160k Transactions Per Second on the Aptos Blockchain](https://medium.com/aptoslabs/block-stm-how-we-execute-over-160k-transactions-per-second-on-the-aptos-blockchain-3b003657e4ba)
[4] [How Sui Works](https://docs.sui.io/devnet/learn/how-sui-works)
[5] [BNB Chain New Milestone: The Implementation of Parallel EVM 2.0](https://www.bnbchain.org/en/blog/new-milestone-the-implementation-of-parallel-evm-2-0/)
[6] [Monad Block-STM Parallel Execution](https://docs.monad.xyz/technical-discussion/execution/parallel-execution)

**Comment by mgarbs** [comment]
Cool idea! Always good to get these new discussions and I like exploring new innovations. A couple of questions for you:

- How will this implementation of parallel transaction execution improve the performance of HSCS over the current implementation? Can you give some examples for clarity?
- Can you provide more information about how the reward and penalty mechanism will work, and what specific incentives are in place for developers to specify correct entity dependencies? Diagrams are super useful with topics like these
- How will the proposed implementation of parallel transaction execution affect the security of HSCS, and what measures would be in place to ensure that transactions are still processed securely?
- Can you provide examples of specific scenarios in which parallel transaction execution would be useful for HSCS users?
- How will the implementation of parallel transaction execution for HSCS differ from other non-EVM chains like Solana, Aptos, and Sui?
- Could you also provide a flow diagram of how this could fit into our existing services?

What you have is great and I think more diagrams addressing the granularity of everything this touches will help drive this from an engineering perspective if they are to implement it.

**Reply by mshakeg** [reply]
@mgarbs thanks for the feedback, I think some of your questions are already addressed in either the initial idea proposal or in the references, but I'll try to clarify further below:

---
> How will this implementation of parallel transaction execution improve the performance of HSCS over the current implementation? Can you give some examples for clarity?

The HSCS uses a single-threaded Besu EVM i.e. all transactions are processed serially. Parallelization can be achieved by multi-threading which could in the best case theoretically increase the HSCS performance by a factor of NUM_THREADS, where NUM_THREADS is the number of threads allocated for the parallel execution of HSCS transactions. Say for example 16 threads were allocated, since each thread is capable of handling 15m gas/s(i.e. current single-threaded gas limit) then in the best case(i.e. all threads are being utilised to their max capacity) a node would effectively be capable of handling 15m gas/s * 16 = 240m gas/s.

---
> Can you provide more information about how the reward and penalty mechanism will work, and what specific incentives are in place for developers to specify correct entity dependencies? Diagrams are super useful with topics like these

I haven't thought too deeply on the effectiveness of the proposed reward and penalty mechanism described in the initial proposal, but I'll expand on it with an example; say a contract tx/call specifies a max gas of `200_000` and a write dependency list of length 2; and say the gas penalty for incorrect(i.e. unused/not called) dependencies set by Hedera is `10_000`, now the penalties are charged pre-emptively(i.e. before the transaction is executed) so `20_000` gas is subtracted from the starting 200k gas leaving 180k gas for execution and say the reward for every correct dependency is `15_000` gas, awarded post-execution.

1. In the case where both of the dependencies specified are incorrect, the transaction incurs a total gas penalty of 20k.
2. In the case where one dependency is incorrect and one is correct the transaction pays a total gas penalty of 10k-15k=-5k; i.e. is still awarded a 5k bonus.
3. In the case where both of the dependencies specified are correct, the transaction is awarded 30k bonus gas

Additionally the bonus can be limited to the min of the bonus gas and a fraction of the starting gas say 20%, so in case 3 the effective bonus gas would be `MIN(30k, 0.2*200k) = 30k`. This bonus gas would be deducted post execution from the gas billing of the transaction(as currently calculated). So for example if the transaction used 40k of the 180k gas available, it would currently be billed for 160k gas(considering the max 20% gas refund on the initial 200k gas), however with the 30k gas bonus it would effectively be billed as if it consumed 130k gas. Similarly if it consumed all 180k gas then with the 30k gas bonus it would effectively be billed as if it consumed 150k gas.

---
> How will the proposed implementation of parallel transaction execution affect the security of HSCS, and what measures would be in place to ensure that transactions are still processed securely?

If implemented properly i.e. contract txs/calls can ONLY call contracts and perform actions on entities/state as defined within the constraints of its dependencies(and revert otherwise) there is no concern of state being corrupted, as transactions are allocated to threads that share at least one dependency in a deterministic fashion. Even if a network/core node has (non-ideally) allocated only a single thread for HSCS evm execution it would still be able to deterministically process the event/"block" of a node that has allocated say N threads(where N>1), albeit slower.

---
> Can you provide examples of specific scenarios in which parallel transaction execution would be useful for HSCS users?

Say you have multiple active dApps(say equal to 16 which is the exact number of threads allocated to the EVM) that submit contract transactions for execution within the same event/"block" of a given core node that all invoke different contracts and specify parallelizable HTS actions i.e. their dependency lists are independent. Now all these dApps don't have to share the same bandwidth(i.e a single thread), but can each have a dedicated thread for processing their respective transactions, so if the average gas cost per transaction across the dApps was 150k gas then all 16 dApps would cumulatively be capable of processing 100 TPS on a single thread at a gas limit of 15m gas/s (or only 6.25 TPS per dApp), however if all transactions were parallelized each dApp would have it's own dedicated bandwidth(i.e. thread) of 15m gas/s which would translate to an average of 100 TPS per dApp. Obviously if you had more dApps than the number of allocated threads then multiple dApps would have to share the same bandwidth(i.e. thread), however the maximum possible gas limit would still significantly higher i.e. gasLimit/thread * NUM_THREADS.

---
> How will the implementation of parallel transaction execution for HSCS differ from other non-EVM chains like Solana, Aptos, and Sui?

It wouldn't be much different from how Solana and Sui implement parallel transaction execution. But it would be different from the Block-STM approach by Aptos. From my limited understanding the Block-STM approach is quite computationally wasteful as transaction's do not specify a dependency list, but instead the network/core node executes transactions in parallel and if say txs A,B,C,... attempted to access dependency/s already accessed by txs X,Y,Z,..., then txs A,B,C,... would be flagged for re-execution and attempted again and the process would repeat until all transactions are processed.

---

**Reply by mgarbs** [reply]
I really like this idea. The only thing I can't get past is how we will maintain the fairness of order for parallel transactions. We have always said that one of the hashgraph's best properties is the fairness of ordering for its transactions. If we change them to be parallel do we compromise that?

**Reply by mshakeg** [reply]
@mgarbs could you please explain how fairness of ordering would be an issue caused by parallel execution? Do the HCS and HTS not already do parallel transaction execution? 

In any case, imu fairness of ordering is not something that can be guaranteed on a node level by the hashgraph algorithm, as each node can change the ordering of transactions within their own events/"blocks" at their own discretion with no consequence(as there's no way for other nodes to determine that re-ordering within an event was done), the hashgraph algorithm only guarantees ordering at the network level, for example, say Alice operates a consensus node and colludes with Mallory to front-run transactions, say to extract MEV from users such as Bob, Bob can mitigate this by submitting his transaction to N nodes say N=2 i.e. Alice's and Carol's nodes, now even if Bob's transaction is moved later in Alice's event/"block"(or even one of Alice's later events) to appear as if it came after Mallory's MEV transaction, the ordering for Bob's transaction will be the median time(as perceived by the supermajority of the network) of the 1st of the N repeated transactions to be processed, which will be increasingly likely to NOT be the transaction in Alice's mal-ordered event as N increases.

And in the case that a node is NOT malicious and does NOT alter the ordering of transactions as it receives them there's still no concern. Say a node receives tx A and then receives tx B that is to be processed within the same event/"block" and tx A and B share at least one dependency then tx B would be allocated to the same thread as tx A to be executed in sequence(and after tx A) within the same thread, so if the node does not go out of its way to re-order transactions then there is no way transactions that share at least one dependency will be executed in a different order other than order the node received them.

**Reply by mgarbs** [reply]
We have a throttle of 15m gas limit that can be lifted to go much higher if we need to. We haven't reached the throttle limit so how can adding parallel transactions be an improvement currently? What is the benefit?

**Reply by mshakeg** [reply]
@mgarbs the HSCS is currently capable of only ~100 TPS(assuming an average gas/tx of 150k and a 15m gas/s limit). Parallelization of the HSCS can result in a higher effective gas limit(a maximum of NUM_THREADS* GAS_LIMIT_PER_THREAD as previously explained) which results in an increased throughput for the HSCS; assuming NUM_THREADS=16 and GAS_LIMIT_PER_THREAD=15m gas/s then a maximum achievable gas limit for the entire HSCS would be 16*15m gas/s = 240m gas/s or a maximum throughput of 1600 TPS(again assuming an average gas/tx of 150k).

**Reply by gregscullard** [reply]
Couple of points I would like to clarify:
* HCS, HTS, Crypto and HSCS transactions are all run in series, they don't operate in separate threads so there is no parallel execution across different services.
* Fairness is guaranteed, if a node maliciously changes the order they process transactions in, that node's signed state will differ from the majority and will no longer be trusted by the network. Furthermore, any tampering with an event's ordering of transactions will break the hash/signature for that event and it will be discarded by other nodes.
* I did discuss parallelisation with Engineering and while the idea is promising (not without its challenges), the feedback at the time was that there are lower hanging fruit that can achieve performance improvements. In other words, parallelising should be done when every other improvement opportunity is exhausted.
* I also understand that greater gas/s is currently possible, the throttles are set to a conservative (and sufficient) level for now and could be increased if/when the need arises.

**Reply by mshakeg** [reply]
@gregscullard thanks for the clarification.

---
> HCS, HTS, Crypto and HSCS transactions are all run in series, they don't operate in separate threads so there is no parallel execution across different services.

Good to know, given that there are already throttles on those services, the throughput if and when parallelization is implemented could possibly be at least an order of magnitude of the current unthrottled throughput, though there may be other bottlenecks such as network bandwidth.

---

> Fairness is guaranteed, if a node maliciously changes the order they process transactions in, that node's signed state will differ from the majority and will no longer be trusted by the network. Furthermore, any tampering with an event's ordering of transactions will break the hash/signature for that event and it will be discarded by other nodes.

I'm not referring to a node retroactively reordering transactions in a previously signed&processed event(which is what I understand from your reply), but re-ordering transactions in an event that is currently being populated/constructed(and that hasn't yet been signed and gossiped). So either the node delays/advances a transaction within the current event that is being populated or even delays the transaction to another of its future events.

---

> I did discuss parallelisation with Engineering and while the idea is promising (not without its challenges), the feedback at the time was that there are lower hanging fruit that can achieve performance improvements. In other words, parallelising should be done when every other improvement opportunity is exhausted.

That's fair, I suspect parallelisation is possibly already being used for some aspects(maybe signature verification) as [network nodes](https://docs.hedera.com/hedera/networks/mainnet/mainnet-nodes/node-requirements) have high CPU requirements(minimum of 48 threads).

---

> I also understand that greater gas/s is currently possible, the throttles are set to a conservative (and sufficient) level for now and could be increased if/when the need arises.

I suspected as much given that the [Besu EVM is capable of 672m gas/block for 2 second blocks](https://www.hyperledger.org/blog/2020/08/06/hyperledger-besu-1-5-performance-enhancements)(i.e. 336m gas/s) in a 5 node network; though Hedera being a larger network will likely not be able to handle such a high single-thread gas limit.

---

**Reply by mgarbs** [reply]
> @mgarbs the HSCS is currently capable of only ~100 TPS(assuming an average gas/tx of 150k and a 15m gas/s limit). Parallelization of the HSCS can result in a higher effective gas limit(a maximum of NUM_THREADS* GAS_LIMIT_PER_THREAD as previously explained) 

Okay, I understand now, thank you @mshakeg. I do think this is a great idea and looks like it could help with throughput of smart contracts. [Looking at metrika over the last 30 days](https://app.metrika.co/hedera/dashboard/network-overview?tr=1M), though, our smart contracts would have to grow a substantial amount to reach the point where it becomes a problem. It's definitely a good topic, thank you for posting it. For me it feels like a better thing to discuss as we approach this limit.

**Reply by mshakeg** [reply]
@mgarbs as @gregscullard mentioned there are other lower hanging fruits to enhance HSCS performance, but parallelization wouldn't be an optimisation limited to only the HSCS but to other services such as the HTS and HCS as well.

**Reply by mgarbs** [reply]
I agree that it can optimize other services too. When we reach our limits, let's take a closer look 👀

**Comment by mshakeg** [comment]
[Sei Network](https://docs.seinetwork.io/smart-contracts-and-local-development/parallelizing-wasm-contracts) employs a much simpler incentive for correctly specifying dependencies i.e. a 50% discount on the gas price. So if all specified dependencies are used then a 50% discount is applied to the tx, however, if at least one incorrect(i.e. unused) dependency is specified then no discount is provided.
